compiler/rustc_trait_selection/src/solve/normalize.rs

```diff
@@ -230,3 +231,42 @@ impl<'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for NormalizationFolder<'_, 'tcx> {
         }
     }
 }
+
+// Deeply normalize a value and return it
+pub(crate) fn deeply_normalize_for_diagnostics<'tcx, T: TypeFoldable<TyCtxt<'tcx>>>(
+    infcx: &InferCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    t: T,
+) -> T {
+    t.clone().fold_with(&mut DeeplyNormalizeForDiagnosticsFolder {
```

Comment:
what's that `clone` doing there? :sweat_smile:

Reply:
Refactoring error